File: components/dvk512/L3G4200D/wiringPi/L3G4200D.c
#include <wiringPi.h>
#include <wiringPiSPI.h>
#include <stdio.h>
#include "L3G4200D.h"

#define channel 0

void L3G4200D_Write(unsigned char add,unsigned char *data,unsigned char len)
{
	unsigned char buf[len+1],i;
	buf[0] = add | 0x40;
	for(i = 0;i < len;i++)
	{
		buf[i+1] = data[i];
	}
	wiringPiSPIDataRW(channel,buf,len+1);
}
void L3G4200D_Read(unsigned char add,unsigned char *data,unsigned char len)
{
	unsigned char buf[len+1],i;
	buf[0] = add | 0xC0;
	for(i = 0;i < len;i++)
	{
		buf[i+1] = 0x11;
	}
    wiringPiSPIDataRW(channel,buf,len+1);
	for(i = 0;i <len;i++)
	{
		data[i] = buf[i+1];
	}
}

void L3G4200D_Init(void)
{
	unsigned char data[5]={0xcf,0x01,0x08,0x00,0x02};
    L3G4200D_Write(CTRL_REG1,data,5);
}
void Read_L3G4200D(float *data)
{
	float Sensitivity;
	unsigned char buf[6];
	L3G4200D_Read(CTRL_REG4,buf,1);
	switch(buf[0]&0x30)
	{
		case 0x00:Sensitivity = 8.75;break;
		case 0x10:Sensitivity = 17.5;break;
		case 0x20:Sensitivity =  70;break;
		case 0x30:Sensitivity =  70;break;
	}
	L3G4200D_Read(OUT_X_L,buf,6);
	data[0] =(short)(buf[0]+((short)buf[1]<<8))*Sensitivity/1000;
	data[1] =(short)(buf[2]+((short)buf[3]<<8))*Sensitivity/1000;
	data[2] =(short)(buf[4]+((short)buf[5]<<8))*Sensitivity/1000;
}

void data_int(float *buf)
{
  	unsigned char i;
	float dat[3];
	for(i=0;i<100;i++)
	{ 
		
		Read_L3G4200D(dat);
		delay(10);
		buf[0]+=dat[0];
		buf[2]+=dat[1];
		buf[1]+=dat[2];
 	 }
  	for(i=0;i<3;i++)
	 	buf[i]=buf[i]/100;
	printf ("%3.3f %3.3f %3.3f\n",buf[0],buf[1],buf[2]);	
}

int main(void)
{
	float data[3]={0.0,0.0,0.0};
	float bet[3]={0.0,0.0,0.0};
	float angle_x=0,angle_y=0,angle_z=0;
	char t;

	if (wiringPiSetup() < 0)return 1;
	wiringPiSPISetup(channel,200000); //2M

	L3G4200D_Init();
	printf("****** L3G4200D *****\n");
	data_int(bet);

	while(1)
	{
		
		Read_L3G4200D(data);
		data[0]=(int)(data[0]-bet[0]);
		data[1]=(int)(data[1]-bet[1]);
		data[2]=(int)(data[2]-bet[2]);    
    
        data[0]=(int)(data[0]); 
        data[1]=(int)(data[1]); 
        data[2]=(int)(data[2]);    
        
		data[0]/=100;
        data[1]/=100;
        data[2]/=100;
       
		angle_x=(data[0])+angle_x;  if(angle_x>=360)angle_x-=360;
		angle_y=(data[1])+angle_y;  if(angle_y>=360)angle_y-=360;
		angle_z=(data[2])+angle_z;  if(angle_z>=360)angle_z-=360;	
		t++;
		if(t==100)t=0;
		if(t==10)
		{
			printf("\nx=%f rad/s\r\n",data[0]); 
        	printf("y=%f rad/s\r\n",data[1]);
        	printf("z=%f rad/s\r\n",data[2]);    
			delay(1000);
		}
	}
}

File: components/dvk512/L3G4200D/python/L3G4200D.py
import spidev
import time

WHO_AM_I = 0x0F
CTRL_REG1 = 0x20
CTRL_REG2 = 0x21
CTRL_REG3 = 0x22
CTRL_REG4 = 0x23
CTRL_REG5 = 0x24
REFERENCE = 0x25
OUT_TEMP = 0x26
STATUS_REG = 0x27
OUT_X_L = 0x28
OUT_X_H = 0x29
OUT_Y_L = 0x2A
OUT_Y_H = 0x2B
OUT_Z_L = 0x2C
OUT_Z_H = 0x2D
FIFO_CTRL_REG = 0x2E
FIFO_SRC_REG  = 0x2F
INT1_CFG = 0x30
INT1_SRC = 0x31
INT1_TSH_XH = 0x32
INT1_TSH_XL = 0x33
INT1_TSH_YH = 0x34
INT1_TSH_YL = 0x35
INT1_TSH_ZH = 0x36
INT1_TSH_ZL = 0x37
INT1_DURATION = 0x38

bus = 0
device = 0
spi= spidev.SpiDev(bus,device)

def L3G4200D_write(add,data=[]):
	data = [add|0x40]+data
	spi.xfer(data)

def L3G4200D_read(add,n):
	data = [add|0xC0]+[0xFF]*n
	data = spi.xfer(data)
	data = data[1:]
	return data

def L3G4200D_init():
	L3G4200D_write(CTRL_REG1,[0xCF,0x01,0x08,0x00,0x02])

def read_L3G4200D():
	data = [0,0,0]
	buf = L3G4200D_read(CTRL_REG4,1)
	fs = buf[0] & 0x30
	if fs  == 0x00:
		Sensitivity = 8.75
	elif fs == 0x10:
		Sensitivity = 17.5
	elif fs == 0x20:
		sensitivity = 70
	else :
		sensitivity = 70
	buf = L3G4200D_read(OUT_X_L,6)
	data[0] = (buf[0] + buf[1] * 0x100)
	data[1] = (buf[2] + buf[3] * 0x100)
	data[2] = (buf[4] + buf[5] * 0x100)
	for i in range(0,3):
		if data[i] > 0x7FFF:
			data[i] -= 0x10000
		data[i] = data[i]*Sensitivity/1000
	return data

def data_int():
	buf = [0,0,0]
	for i in range(0,100):
		data = read_L3G4200D()
		time.sleep(0.01)
		buf[0] += data[0]
		buf[1] += data[1]
		buf[2] += data[2]
	for i in range(0,3):
		buf[i] = buf[i]/100
	print("%3.3f %3.3f %3.3f\n" %(buf[0],buf[1],buf[2]))
	return buf

print("****** L3G4200D ******")
L3G4200D_init()
bet = data_int() 

while True:
	data = read_L3G4200D()
	data[0] = (data[0]-bet[0])/100
	data[1] = (data[1]-bet[1])/100
	data[2] = (data[2]-bet[2])/100
	
	print("\nx=%.3f rad/s" %data[0]) 
	print("y=%.3f rad/s" %data[1])
	print("z=%.3f rad/s" %data[2])
	time.sleep(1)
